File: implementations/dart/lib/src/i57.dart
library b57_i57;

import 'dart:math';
import 'b57.dart' as b57;
import 'h57.dart';
import 'id57.dart';

String i57Encode(List<int> input) => b57.encode(input);
List<int> i57Decode(String input) => b57.decode(input);

String i57Hash(List<int> input, H57Length length) =>
    h57Hash(input, length);

String i57Random(int bytes) {
  final rand = Random.secure();
  final data = List<int>.generate(bytes, (_) => rand.nextInt(256));
  return b57.encode(data);
}

String i57Id(List<int> input, [int length = ID57Length.def]) =>
    id57Generate(input, length);

bool i57IsValid(String input) => b57.isValid(input);
bool i57IsCanonical(String input) => b57.isCanonical(input);

/// i57_validate_identifier (I57 Core API 5.3).
///
/// length_enum's SIGN selects the check, mirroring ID57 Core API 7.3/
/// 11.4/11.5:
///
/// - negative (FIXED width): delegates entirely to [id57IsLength] - valid
///   B57 AND exact length == -length. MUST NOT decode/canonicalize (fixed
///   widths are bignum prefixes, not canonical B57).
/// - positive/DEFAULT (bit length): [id57IsLength] does not apply, so the
///   bound + canonical + decoded-byte-length + excess-bits check is
///   implemented directly here.
bool i57ValidateIdentifier(String input, [int length = ID57Length.def]) {
  final effective = resolveID57Length(length);

  if (effective < 0) {
    return id57IsLength(input, effective);
  }

  final bits = effective;
  final byteLength = (bits + 7) ~/ 8;
  final range = id57Range(effective);

  if (input.length < range.min || input.length > range.max) {
    return false;
  }
  if (!id57IsCanonical(input)) {
    return false;
  }

  final decoded = b57.decode(input);
  if (decoded.length != byteLength) {
    return false;
  }

  final excessBits = byteLength * 8 - bits;
  if (excessBits > 0) {
    final lastByte = decoded[byteLength - 1];
    final mask = (1 << excessBits) - 1;
    if ((lastByte & mask) != 0) {
      return false;
    }
  }

  return true;
}

File: implementations/go/i57.go
package b57

import "strings"

// I57Encode encodes arbitrary bytes to a string using B57.
// It proxies directly to Encode.
func I57Encode(input []byte) string {
	return Encode(input)
}

// I57Decode decodes a B57 string back to bytes.
// It proxies directly to Decode.
func I57Decode(input string) ([]byte, error) {
	return Decode(input)
}

// I57Hash generates a B57-encoded hash using the specified hash function and length.
// It proxies directly to H57Hash.
func I57Hash(input []byte, length H57Length) (string, error) {
	return H57Hash(input, length)
}

// I57Random generates a random B57 string according to the specified mode.
// It proxies directly to R57Generate.
func I57Random(mode R57Mode) (string, error) {
	return R57Generate(mode)
}

// I57Id generates an identifier by hashing the input and truncating to the specified ID length.
// It proxies directly to ID57Generate.
func I57Id(input []byte, length ID57Length) (string, error) {
	return ID57Generate(input, length)
}

// I57IsValid checks if a string is a valid B57 value under integration-layer rules.
// The integration layer rejects empty strings by default.
func I57IsValid(input string) bool {
	if input == "" {
		return false
	}
	return IsValid(input)
}

// I57IsCanonical checks if a string is canonically encoded in B57.
func I57IsCanonical(input string) bool {
	if input == "" {
		return false
	}
	return IsCanonical(input)
}

// I57ValidateIdentifier validates input as an ID57 identifier produced for
// length_enum, dispatching on the sign of length_enum (i57-core-api.txt 5.3):
//
//   - length_enum < 0 (FIXED width): delegates entirely to ID57IsLength
//     (valid B57 alphabet AND exact length == -length_enum). The string
//     MUST NOT be decoded/canonicalized - fixed-width output is a prefix
//     of a bignum encoding, not canonical B57.
//   - length_enum >= 0 (BIT length, including ID57Default): ID57IsLength
//     does not apply, so the bound + canonical + decoded-byte-length +
//     excess-bit check is implemented directly here.
func I57ValidateIdentifier(input string, length ID57Length) bool {
	effective, err := resolveID57Length(length)
	if err != nil {
		return false
	}

	if effective < 0 {
		ok, err := ID57IsLength(input, effective)
		if err != nil {
			return false
		}
		return ok
	}

	minChars, maxChars, err := ID57Range(effective)
	if err != nil {
		return false
	}
	if len(input) < minChars || len(input) > maxChars {
		return false
	}
	if !ID57IsCanonical(input) {
		return false
	}

	decoded, err := Decode(input)
	if err != nil {
		return false
	}

	bits := id57BitsByLength[effective]
	byteLength := (bits + 7) / 8
	if len(decoded) != byteLength {
		return false
	}

	excessBits := byteLength*8 - bits
	if excessBits > 0 {
		lastByte := decoded[byteLength-1]
		mask := byte((1 << excessBits) - 1)
		if lastByte&mask != 0 {
			return false
		}
	}

	return true
}

// I57ValidateEntropy performs a lightweight heuristic check for obviously
// low-entropy-looking identifiers. It must not be used for security decisions.
func I57ValidateEntropy(input string) bool {
	if !I57ValidateIdentifier(input, ID57Default) {
		return false
	}
	if !passesCharacterDiversity(input) {
		return false
	}
	if hasRepeatedHalfPattern(input) {
		return false
	}
	if isSingleRepeatedPattern(strings.ToLower(input)) {
		return false
	}

	return true
}

func passesCharacterDiversity(input string) bool {
	unique := make(map[byte]struct{}, len(input))
	maxRun := 1
	run := 1

	for i := 0; i < len(input); i++ {
		unique[input[i]] = struct{}{}
		if i > 0 {
			if input[i] == input[i-1] {
				run++
				if run > maxRun {
					maxRun = run
				}
			} else {
				run = 1
			}
		}
	}

	if len(unique) < 4 {
		return false
	}
	if maxRun > len(input)/2 {
		return false
	}
	return true
}

func hasRepeatedHalfPattern(input string) bool {
	if len(input)%2 != 0 {
		return false
	}
	halfLen := len(input) / 2
	firstHalf := input[:halfLen]
	secondHalf := input[halfLen:]
	return firstHalf == secondHalf
}

func isSingleRepeatedPattern(input string) bool {
	if input == "" {
		return false
	}
	return input == strings.Repeat(string(input[0]), len(input))
}
